avr-hal-generic/src/wdt.rs

```diff
@@ -0,0 +1,141 @@
+//! WDT Implementation
+
+pub use embedded_hal::watchdog;
+
+/// Approximate length of the time-out period before the watchdog provides a system reset.
+///
+/// After enabling the watchdog timer, call [`feed`](trait.Watchdog.html#tymethod.feed) before the
+/// period ends to prevent a reset.
+///
+/// Some AVR models do not support all period lengths.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum WatchdogTimeOutPeriod {
+    /// 16 milliseconds
+    Ms16,
+    /// 32 milliseconds
+    Ms32,
+    /// 64 milliseconds
+    Ms64,
+    /// 125 milliseconds
+    Ms125,
+    /// 250 milliseconds
+    Ms250,
+    /// 500 milliseconds
+    Ms500,
+    /// 1 second
+    Ms1000,
+    /// 2 seconds
+    Ms2000,
+    /// 4 seconds
+    Ms4000,
+    /// 8 seconds
+    Ms8000,
+}
+
+/// Implement traits for the watchdog interface
+#[macro_export]
+macro_rules! impl_wdt {
+    (
+        $(#[$wdt_attr:meta])*
+        pub struct $Wdt:ident {
+            mcu_status_register: $MCUSR:ty,
+            peripheral: $WDT:ty,
+        }
+    ) => {
+        use avr_device::generic::Reg;
+        pub use avr_hal::hal::watchdog::*;
+
+        pub struct $Wdt {
+            peripheral: $WDT,
+        }
+
+        impl $Wdt {
+            /// Initializes a Wdt.
+            ///
+            /// If a prior reset was provided by the watchdog, the WDRF in MCUSR would be set, so
+            /// WDRF is also cleared to allow for re-enabling the watchdog.
+            pub fn new(mcu_status_register: &Reg<$MCUSR>, peripheral: $WDT) -> Self {
```

Comment:
I would suggest using the type alias from `avr-device` instead of naming `Reg<>` here.  That means, I'd refer to [`avr_device::atmega328p::cpu::MCUSR`](https://docs.rs/avr-device/0.2.2/avr_device/atmega328p/cpu/type.MCUSR.html).

```suggestion
            pub fn new(mcu_status_register: &$MCUSR, peripheral: $WDT) -> Self {
```